### PR TITLE
[Doc] Guide: how to write .NET Application

### PR DESCRIPTION
**Pull request by again4you**
This patch newly adds the guide for writing .NET Application with
Machine Learning APIs.

Signed-off-by: Sangjung Woo <sangjung.woo@samsung.com>

**Self evaluation:**
1. Build test: [ ]Passed [ ]Failed [*]Skipped
2. Run test: [ ]Passed [ ]Failed [* ]Skipped

#### Submit v3
* Apply all comments from reviews
* Sync from Tizen official documentation

### CODE REVIEW

File: Documentation/writing-tizen-csharp-apps.md [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+## Installing Visual Studio Tools for Tizen
+In order to use this functionality, you need to install Visual Studio Tools for Tizen and Tizen SDK version 5.5 or higher. You can find the detailed guide for this at the below link.
+
+* https://developer.tizen.org/development/visual-studio-tools-tizen/installing-visual-studio-tools-tizen
+
+
+# Machine Learning Inference
+
+The Machine Learning Inference features introduce how you can easily invoke the neural network model and get the the inference output result.
```
Comment (dongju-chae) [inline Documentation/writing-tizen-csharp-apps.md]: There are same typos: `the the`

File: Documentation/writing-tizen-csharp-apps.md [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+TensorsInfo in_info = new TensorsInfo();
+in_info.AddTensorInfo(TensorType.UInt8, new int[4] { 3, 224, 224, 1 });
+
+/* Output Dimension: 1001 for classification */
+TensorsData out_data = new TensorsInfo();
+out_info.AddTensorInfo(TensorType.UInt8, new int[4] { 1001, 1, 1, 1 });
+```
+
+<a name="load"></a>
+## Loading the neural network model and configures an runtime environment
```
Comment (dongju-chae) [inline Documentation/writing-tizen-csharp-apps.md]: an runtime --> a runtime

File: Documentation/writing-tizen-csharp-apps.md [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+```C#
+/* Create SingleShot instance with model information */
+SingleShot single = new SingleShot(model_path, in_info, out_info);
+```
+
+<a name="invoke"></a>
+## Invoking the neural network model with the given input data
+
+In order to invoke the model, you need to make the `Tizen.MachineLearning.Inference.TensorsData` instance to pass its input data. You can adds the various types of Tensor data, which are already specified at `TensorInfo` instance. However, the maximum size of the TensorsData is 16. `IndexOutOfRangeException` could raise when exceeding the limit. Input data is passed as a byte array format (i.e. byte[]).
```
Comment (dongju-chae) [inline Documentation/writing-tizen-csharp-apps.md]: can adds --> can add

File: Documentation/writing-tizen-csharp-apps.md (hunk 1 above)
Comment (again4you) [reply Documentation/writing-tizen-csharp-apps.md]: ACK!

File: Documentation/writing-tizen-csharp-apps.md (hunk 2 above)
Comment (again4you) [reply Documentation/writing-tizen-csharp-apps.md]: ACK!

File: Documentation/writing-tizen-csharp-apps.md (hunk 3 above)
Comment (again4you) [reply Documentation/writing-tizen-csharp-apps.md]: ACK!

**Comment by again4you** [comment]
## Submit v2
* Fix the typo.

File: Documentation/writing-tizen-csharp-apps.md [hunk 4]
```diff
@@ -0,0 +1,131 @@
+# Writing a Tizen .NET Application
+
+This document provides you how to write a Tizen .NET Application with Machine Learning APIs.
+Since Tizen 5.5 (not released yet), Machine Learning Inference functionality has been provided on Mobile, Wearable and TV profile.
```
Comment (myungjoo) [inline Documentation/writing-tizen-csharp-apps.md]: Remove ```(not released yet)```. Please assume that it's already released. We don't want to update docs with Tizen releases.

File: Documentation/writing-tizen-csharp-apps.md [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+## Installing Visual Studio Tools for Tizen
+In order to use this functionality, you need to install Visual Studio Tools for Tizen and Tizen SDK version 5.5 or higher. You can find the detailed guide for this at the below link.
```
Comment (myungjoo) [inline Documentation/writing-tizen-csharp-apps.md]: There is no Tizen SDK. It's Tizen Studio.
Besides, there won't be Tizen Studio 5.5 in 2019 or 2020. The most recent version of Tizen studio as of 2019-10-18 is 3.3 (Studio version goes independently from Tizen version.)

File: Documentation/writing-tizen-csharp-apps.md [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+
+* https://developer.tizen.org/development/visual-studio-tools-tizen/installing-visual-studio-tools-tizen
+
+
+# Machine Learning Inference
+
+The Machine Learning Inference features introduce how you can easily invoke the neural network model and get the inference output result.
```
Comment (myungjoo) [inline Documentation/writing-tizen-csharp-apps.md]: Rephrase:
```
The Machine Learning Inference feature allows developers to invoke neural network models with less efforts and higher efficiency.
```

File: Documentation/writing-tizen-csharp-apps.md [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+With Tizen.MachineLearning.Inference.SingleShot, you can load the existing neural network model or your own specific model from storage and invoke it with a single instance of input data, finally get the inference output result.
+
+Currently, this is an experimental functionality and is likely to be changed.
```
Comment (myungjoo) [inline Documentation/writing-tizen-csharp-apps.md]: Is it really experimental?

File: Documentation/writing-tizen-csharp-apps.md [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+
+The main features of the Tizen.MachineLearning.Inference are:
+
+- Managing a Tensor information
```
Comment (myungjoo) [inline Documentation/writing-tizen-csharp-apps.md]: information is not countable.

File: Documentation/writing-tizen-csharp-apps.md [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+
+  You can [configure the input and output Tensor Information](#manage) such as its name, data type and dimension.
+
+- Loading the neural network model and configures a runtime environment
+
+  You can [load the neural network model from storage and configure a runtime environment](#load).
+
+- Invoking the neural network model with the input data
```
Comment (myungjoo) [inline Documentation/writing-tizen-csharp-apps.md]: Remove ```the``` from ```the input data```.

File: Documentation/writing-tizen-csharp-apps.md [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+
+Currently, this is an experimental functionality and is likely to be changed.
+
+The main features of the Tizen.MachineLearning.Inference are:
+
+- Managing a Tensor information
+
+  You can [configure the input and output Tensor Information](#manage) such as its name, data type and dimension.
+
+- Loading the neural network model and configures a runtime environment
```
Comment (myungjoo) [inline Documentation/writing-tizen-csharp-apps.md]: ```the neural network model``` --> ```a neural network model```

Comment (myungjoo) [reply Documentation/writing-tizen-csharp-apps.md]: ```configures``` --> ```configuring```

File: Documentation/writing-tizen-csharp-apps.md [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+
+  After setting up the SingleShot instance with its required information, you can [invoke the model with the input data and get the inference output result](#invoke).
+
+- Fetching the inference result after invoking
+
+  You can [fetch the inference result after invoking](#fetch).
+
+## Prerequisites
+
+To enable your application to use the Machine Learning Inference functionality:
```
Comment (myungjoo) [inline Documentation/writing-tizen-csharp-apps.md]: To use Machine Learning Inference with your application:

File: Documentation/writing-tizen-csharp-apps.md [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+      <!--To access media storage-->
+      <privilege>http://tizen.org/privilege/mediastorage</privilege>
+
+      <!--To access, read, and write to the external storage-->
+      <privilege>http://tizen.org/privilege/externalstorage</privilege>
+    </privileges>
+    ```
+
+<a name="manage"></a>
+## Managing a Tensor information
```
Comment (myungjoo) [inline Documentation/writing-tizen-csharp-apps.md]: remove a

File: Documentation/writing-tizen-csharp-apps.md [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+
+In this example, we are going to use the MobileNet v1 model for TensorFlow Lite, which is used for image classification. Its input data type is `Unsigned Integer` and its input dimension is `3 X 224 X 224`. Its output data type is same but its output dimension is `1001 X 1 X 1 X 1`.
```
Comment (myungjoo) [inline Documentation/writing-tizen-csharp-apps.md]: ```we are going to use``` --> ```we use```
```model for TensorFlow Lite``` --> ```model with TensorFlow Lite```
```which is used for``` --> ```which does```

File: Documentation/writing-tizen-csharp-apps.md [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+After calling `Invoke()` method of `Tizen.MachineLearning.Inference.SingleShot` class, 
+`Tizen.MachineLearning.Inference.TensorsData` instance is returned as the inference result.
+It can have multiple output data so you have to fetch each data by using `GetTensorData()` method. `IndexOutOfRangeException` could raise when exceeding the limit.
+
+```C#
+/* Get the first Tensor data from the inference result */
+byte[] out_buffer = out_data.GetTensorData(0);
+```
+
+As I mentioned [above](#invoke), `TensorsData` class is used to send the Input data to a neural network model. Moreover, it provides the `Count` property for developers to get the number or Tensor it has.
```
Comment (myungjoo) [inline Documentation/writing-tizen-csharp-apps.md]: Use we instead of I.

File: Documentation/writing-tizen-csharp-apps.md (hunk 4 above)
Comment (again4you) [reply Documentation/writing-tizen-csharp-apps.md]: ACK!

File: Documentation/writing-tizen-csharp-apps.md (hunk 5 above)
Comment (again4you) [reply Documentation/writing-tizen-csharp-apps.md]: According to the [official guide](https://developer.tizen.org/development/visual-studio-tools-tizen/installing-visual-studio-tools-tizen#install) for `Installing Visual Studio Tools for Tizen`, developers should install both 'Tools for Tizen' and 'Tizen Baseline SDK'. I found out that there is no specific version for SDK. So I will fix it as below.

> In order to use this functionality, you need to install Visual Studio Tools for Tizen and Tizen SDK. You can find the detailed guide for this at the below link.

File: Documentation/writing-tizen-csharp-apps.md (hunk 7 above)
Comment (again4you) [reply Documentation/writing-tizen-csharp-apps.md]: ACK. I will remove an `experimental functionality` and add your sentence.

File: Documentation/writing-tizen-csharp-apps.md (hunk 8 above)
Comment (again4you) [reply Documentation/writing-tizen-csharp-apps.md]: ACK!

File: Documentation/writing-tizen-csharp-apps.md (hunk 9 above)
Comment (again4you) [reply Documentation/writing-tizen-csharp-apps.md]: ACK!

File: Documentation/writing-tizen-csharp-apps.md (hunk 10 above)
Comment (again4you) [reply Documentation/writing-tizen-csharp-apps.md]: ACK!

File: Documentation/writing-tizen-csharp-apps.md (hunk 12 above)
Comment (again4you) [reply Documentation/writing-tizen-csharp-apps.md]: ACK!

File: Documentation/writing-tizen-csharp-apps.md (hunk 11 above)
Comment (again4you) [reply Documentation/writing-tizen-csharp-apps.md]: ACK!

File: Documentation/writing-tizen-csharp-apps.md (hunk 13 above)
Comment (again4you) [reply Documentation/writing-tizen-csharp-apps.md]: ACK!

File: Documentation/writing-tizen-csharp-apps.md (hunk 14 above)
Comment (again4you) [reply Documentation/writing-tizen-csharp-apps.md]: ACK!

File: Documentation/writing-tizen-csharp-apps.md (hunk 6 above)
Comment (again4you) [reply Documentation/writing-tizen-csharp-apps.md]: ACK!

**Comment by again4you** [comment]
### Submit v3
* Apply myungjoo's comments

**Comment by myungjoo** [comment]
This is same with https://github.com/Samsung/tizen-docs/pull/891 , right?
I'll wait for merging of https://github.com/Samsung/tizen-docs/pull/891 and then, I'll merge this after you apply changes from 891.

File: Documentation/writing-tizen-csharp-apps.md [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+After calling `Invoke()` method of `Tizen.MachineLearning.Inference.SingleShot` class, 
+`Tizen.MachineLearning.Inference.TensorsData` instance is returned as the inference result.
+It can have multiple output data so you have to fetch each data by using `GetTensorData()` method. `IndexOutOfRangeException` could raise when exceeding the limit.
+
+```C#
+/* Get the first Tensor data from the inference result */
+byte[] out_buffer = out_data.GetTensorData(0);
+```
+
+As I have mentioned [above](#invoke), `TensorsData` class is used to send the Input data to a neural network model. Moreover, it provides the `Count` property for developers to get the number or Tensor it has.
```
Comment (wooksong) [inline Documentation/writing-tizen-csharp-apps.md]: I am not sure about 'I' here. How about just 

> As mentioned [above](#invoke),

Comment (again4you) [reply Documentation/writing-tizen-csharp-apps.md]: ACK!

File: Documentation/writing-tizen-csharp-apps.md [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+
+# Machine Learning Inference
+
+The Machine Learning Inference feature allows developers to invoke neural network models with less efforts and higher efficiency.
+
+With Tizen.MachineLearning.Inference.SingleShot, you can load the existing neural network model or your own specific model from storage and invoke it with a single instance of input data, finally get the inference output result.
+
+The feature of Tizen.MarhineLearning.Inference.Pipeline allows to manage the topology of
+input and output data streams and interconnections between processors and
+neural network models with simple API calls. However, this feature is available with C-API in Tizen 5.5 and not available with .NET in Tizen 5.5. This feature will be soon available for .NET with the next version of Tizen.
```
[the remark this answers is not in the document]
Comment (again4you) [reply Documentation/writing-tizen-csharp-apps.md]: Maybe next version of Tizen. I hope.

File: Documentation/writing-tizen-csharp-apps.md [hunk 17]
```diff
@@ -0,0 +1,133 @@
+# Writing a Tizen .NET Application
+
+This document provides you how to write a Tizen .NET Application with Machine Learning APIs.
```
[the remark this answers is not in the document]
Comment (again4you) [reply Documentation/writing-tizen-csharp-apps.md]: Right. I will.

**Comment by again4you** [comment]
#### Submit v3
* Apply all comments from reviews
* Sync from Tizen official documentation

File: Documentation/writing-tizen-csharp-apps.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (wooksong) [reply Documentation/writing-tizen-csharp-apps.md]: @again4you I mean
s/will be soon available/will be available soon

**Pull request merged**